enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+
+	for each rule in new_obj.egress:
+		if rule.to.dnsName != "" && check_dns_name_used_in_egress_firewall(rule.to.dnsName, old_obj) == false
+			&& check_dns_name_resolver_exists(rule.to.dnsName) == false:
+			create_dns_name_resolver(rule.to.dnsName)
+
+else if match(event.type, "delete"):
+
+	for each rule in obj.egress:
+		if rule.to.dnsName != "" && check_dns_name_used_in_any_egress_firewall(rule.to.dnsName) == false:
+			delete_dns_name_resolver(rule.to.dnsName)
+```
+
+* The OVN-K cluster manager will watch for events related to `EgressFirewall` objects.
+* When an OpenShift cluster administrator creates/updates an EgressFirewall resource for a namespace and adds rule(s) containing DNS name(s),
+the cluster manager will receive the create/update event.
+* The OVN-K cluster manager will create corresponding `DNSNameResolver` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the `openshift-ovn-kubernetes` namespace. The name of the CR will be assigned using a hash
+function (similar to the ComputeHash [here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
+prefixed by `dns-`. The input to the hash function will be the DNS name. The `.spec.name` field of the CR will be set to the DNS name along with a trailing period.
+The trailing period is added because DNS servers store DNS names with a trailing period. Thus, to make it consistent, and also for ease of matching,
+the `spec.name` will be set with a trailing period.
+* When an OpenShift cluster administrator deletes an EgressFirewall resource for a namespace containing rule(s) for DNS name(s)
+OR updates an EgressFirewall resource for a namespace and deletes rule(s) containing DNS name(s), the cluster manager will receive the delete/update event.
+* The cluster manager will then check if the same DNS names are also used in the EgressFirewall rules in other namespaces. If the
+DNS names are not used, then the cluster manager will delete the corresponding `DNSNameResolver` CRs.
+
+#### Workflow of `DNSNameResolver` controller
+
+The following pseudocode explains the workflow of the `DNSNameResolver` controller:
+```shell
+watch(DNSNameResolver)
+run_dns_resolution_in_background()
+
+if match(event.type, "create") || match(event.type, "update"):
+
+	if check_dns_name_added(obj.spec.name) == false:
+		add_and_resolve_dns_name(obj.spec.name)
+
+	for each resolvedName in obj.status.resolvedNames:
+		ensure_dns_name_added(resolvedName)
+	
+	if removal_of_ips_required(obj):
+		update_dns_resolver(obj)
+		return
+
+	remaining_time_till_grace_period, is_expiring = grace_period_expires_after(obj)
+
+	if is_expiring:
+		reconcile_event_after(remaining_time_till_grace_period)
+		return
+
+else if match(event.type, "delete"):
+
+	if is_wildcard(obj.spec.name) == true
+		|| (is_wildcard(obj.spec.name) == false && match_wildcard_dns_name(obj.spec.name) == false):
```

Comment:
What happens if one of the names of a wildcard is subsequently added to the deny list?  E.g. we allowed *.example.com, then someone gained access to badactor.example.com and so we added badactor.example.com to the deny list?

Reply:
Only allow rules will be supported for DNS based EgressFirewall rules: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R36-R40

It was suggested here by @danwinship: https://github.com/openshift/enhancements/pull/1335#discussion_r1101797917